### PR TITLE
[ENH] Add Sparse Matrix Support For `HistGradientBoostingClassifier`

### LINKED ISSUE #12: [DEV] (#15336) Add Sparse Matrix Support For `HistGradientBoostingClassifier`
Implement the fix/feature and submit a set of tests, described in a form that would allow any user to download your new version of a scikit-learn component, run it, execute the tests, and determine that the software works correctly. 

You can assume that the user has a basic working knowledge of Python and scikit-learn. Write a brief technical commentary on how your changes affect the design and/or the code of scikit-learn. List all relevant scikit-learn source code files that were added, modified or removed as part of your implementation work.

**Task List**:

- [x] Submit PR to [scikit-learn/scikit-learn](https://github.com/scikit-learn/scikit-learn) or our group fork [navn-r/scikit-learn](https://github.com/navn-r/scikit-learn). Include appropriate code changes and unit tests

### PR DESCRIPTION
**Pull request by navn-r**
#### Reference Issues/PRs

Fixes: #12 

#### What does this implement/fix? Explain your changes.

- Adds Sparse Matrix Support For `HistGradientBoostingClassifier`

**TODO**:
- [x] Add comment explanations (see https://github.com/navn-r/scikit-learn/pull/14#pullrequestreview-929721014)
- [x] Add or modify unit tests for the affected area

### CODE REVIEW

File: sklearn/ensemble/_hist_gradient_boosting/binning.py [hunk 1]
```diff
@@ -19,6 +19,76 @@
...
+    col_lengths = [len(x) for x in columns]
+    max_length_col = max(col_lengths)
+
+    for i, col_length in enumerate(col_lengths):
+        column, index = columns[i], indices[i]
+
+        column_dist = max_length_col - col_length
+        index_dist = max_length_col - len(index)
+
+        # TODO @D01 ADD Explanation why here
```
Comment (navn-r) [inline sklearn/ensemble/_hist_gradient_boosting/binning.py]:
```suggestion
```

File: sklearn/ensemble/_hist_gradient_boosting/binning.py [hunk 2]
```diff
@@ -41,12 +111,25 @@ def _find_binning_thresholds(col_data, max_bins):
...
+    if not sp.issparse(col_data):
+        # ignore missing values when computing bin thresholds
+        missing_mask = np.isnan(col_data)
+        if missing_mask.any():
+            col_data = col_data[~missing_mask]
+        col_data = np.ascontiguousarray(col_data, dtype=X_DTYPE)
+        distinct_values = np.unique(col_data)
+    else:
+        # ignore zeros when computing distinct values
+        # TODO @D01 Explain how `col_data[col_data.nonzero()[0]]` works
```
Comment (navn-r) [inline sklearn/ensemble/_hist_gradient_boosting/binning.py]:
```suggestion
```

**Review by navn-r — COMMENTED** [review]
Add comment explanations  :)

**Review by Samyakk123 — APPROVED** [review]
LGTM!

**Review by tapasrastogi2411 — APPROVED** [review]
LGTM! :)

**Review by divyam-p — APPROVED** [review]
Looks good 👍

**Pull request merged**